### PR TITLE
Int-construction

### PR DESCRIPTION
**Pull request by toumix**
This PR implements the Int construction, see https://docs.discopy.org/en/int/_api/discopy.interaction.html

The main new feature is a new `interaction` module which implements the free compact category generated by a traced symmetric category, or more generally the free ribbon category generated by a traced balanced category.
This includes:

* a `Ty` class defined by a pair `positive` and `negative` of some type `Ty.natural`
* a `Diagram` class defined by an `inside` of type `Diagram.natural`, `dom` and `cod`
* methods for identity, composition, tensor, braid, cups and caps of interaction diagrams
* a method `interaction.Diagram.simplify` which goes back and forth to `hypergraph.Diagram`

This required a change to the `hypergraph` module:

* `upgrade` and `downgrade` take new optional factory arguments
* `make_progressive` now introduces `Trace` boxes rather than `Cup` and `Cap`

### CODE REVIEW

File: discopy/utils.py [hunk 1]
```diff
@@ -253,6 +313,7 @@ def assert_isinstance(object, cls: type | tuple[type, ...]):
 
 def mmap(binary_method):
     """ Turn a binary method into n-ary. """
```
Comment (boldar99) [inline discopy/utils.py:315]: This description is quite misleading. I thought it'd change the binary input (binary number) into some other basis. Though this is not part of this PR.

File: discopy/utils.py [hunk 2]
```diff
@@ -82,6 +90,58 @@ def then(self, other: MappingOrCallable[VT, V2T]
...
+    >>> from dataclasses import dataclass
+    >>> @dataclass
+    ... class L(NamedGeneric["type_param"]):
+    ...     inside: list
+    >>> assert L[int]([1, 2, 3]).type_param == int
+    >>> assert L[int]([1, 2, 3]) != L[float]([1, 2, 3])
+    """
+    def __class_getitem__(self, attr):
+        class Result(Generic[TypeVar(attr)]):
+            def __class_getitem__(cls, dtype: type, _cache=dict()):
```
Comment (boldar99) [inline discopy/utils.py:124]: Cool way to cache! :D

File: docs/discopy.bib [hunk 3]
```diff
@@ -1,3 +1,16 @@
+@inproceedings{Abramsky96,
```
[the remark this answers is not in the document]
Comment (toumix) [reply docs/discopy.bib:1]: Sure, I just added the arXiv handle.

File: discopy/utils.py (hunk 1 above)
Comment (toumix) [reply discopy/utils.py]: I agree this was not very clear. I renamed `mmap` to `unbiased` and added [a link to the nLab](https://ncatlab.org/nlab/show/biased+definition).

File: discopy/utils.py (hunk 2 above)
Comment (toumix) [reply discopy/utils.py:124]: This is described in the official Python guide as a "gotcha that isn't a gotcha" https://docs.python-guide.org/writing/gotchas/#when-the-gotcha-isn-t-a-gotcha

**Review by boldar99 — APPROVED** [review]
I'm not saying that I understand everything that's happening here, but each individual peace seems good, and there are no obvious issues.
